packages/sdk/src/transactions/PSBTBuilder.ts

```diff
@@ -0,0 +1,275 @@
+    const existingInputHashes = this.psbt.txInputs.map((input) => {
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore
+      const hash = reverseBuffer(input.hash) as Buffer
+
+      return generateTxUniqueIdentifier(hash.toString("hex"), input.index)
+    })
+
+    for (const [index, input] of this.inputs.entries()) {
+      if (existingInputHashes.includes(generateTxUniqueIdentifier(input.hash, input.index))) continue
+
+      this.psbt.addInput(input)
+      this.psbt.setInputSequence(index, this.getInputSequence())
+    }
+  }
+
+  private validateOutputAmount() {
+    if (this.outputAmount < MINIMUM_AMOUNT_IN_SATS) {
+      throw new Error(`Output amount too low. Minimum output amount needs to be ${MINIMUM_AMOUNT_IN_SATS} sats`)
+    }
+  }
+
+  private addOutputs() {
+    this.outputs.forEach((output) =>
+      this.psbt.addOutput({
+        address: output.address,
+        value: output.value
+      })
+    )
+  }
+
+  private adjustChangeOutput() {
+    const changeOutput = this.outputs[this.changeOutputIndex]
+    this.outputs[this.changeOutputIndex] = {
+      ...changeOutput,
+      value: this.changeAmount
+    }
+  }
+
+  private removeOutputsByIndex(indexes: number[] = []) {
+    this.outputs = this.outputs.filter((_, index) => !indexes.includes(index))
+  }
+
+  private removeChangeOutput() {
+    this.changeOutputIndex > -1 && this.removeOutputsByIndex([this.changeOutputIndex])
+    this.changeOutputIndex = -1
+  }
+
+  protected async addChangeOutput() {
+    await this.isNegativeChange()
+
+    if (this.changeAmount < MINIMUM_AMOUNT_IN_SATS) {
+      this.removeChangeOutput()
+      return
+    }
+
+    if (this.changeOutputIndex > -1) {
+      return this.adjustChangeOutput()
+    }
+
+    this.outputs.push({
+      address: this.changeAddress || this.address,
+      value: this.changeAmount
+    })
+
+    this.changeOutputIndex = this.outputs.length - 1
+
+    this.calculateChangeAmount()
+  }
+
+  private calculateOutputAmount() {
+    this.outputAmount = Math.floor(this.outputs.reduce((acc, curr) => (acc += curr.value), 0))
+
+    this.validateOutputAmount()
+  }
+
+  private async calculateChangeAmount() {
+    if (this.inscriberMode) return
+
+    this.changeAmount = Math.floor(this.inputAmount - this.outputAmount - this.fee)
+    await this.addChangeOutput()
+  }
+
+  private async isNegativeChange() {
+    if (this.changeAmount >= 0) return
+
+    await this.prepare()
+    if (this.noMoreUTXOS) {
+      throw new Error(`Insufficient balance. Decrease the output amount by ${this.changeAmount * -1} sats`)
+    }
+  }
+
+  private getReservedUTXOs() {
+    return this.utxos.map((utxo) => generateTxUniqueIdentifier(utxo.txid, utxo.n))
+  }
+
+  private async retrieveUTXOs(address?: string, amount?: number) {
+    if (this.inscriberMode && !address) return
+
+    amount = amount && amount > 0 ? amount : this.changeAmount < 0 ? this.changeAmount * -1 : this.outputAmount
```

Comment:
can we add comment as to what this nested ternary condition is achieving? seems like it has alot of amount to fallback to 🤔  can be confusing as the amount changes based on where the function is called.